Avoidable hangs in WebCore::defaultWebCryptoMasterKey

darinadler opened the pull request:
#### 4f8eecb565eb2344d6e27d89b207e299b7f7cdf8
<pre>
Avoidable hangs in WebCore::defaultWebCryptoMasterKey
<a href="https://bugs.webkit.org/show_bug.cgi?id=295844">https://bugs.webkit.org/show_bug.cgi?id=295844</a>
<a href="https://rdar.apple.com/151058747">rdar://151058747</a>

Reviewed by Sam Weinig.

The code in WebCore::defaultWebCryptoMasterKey does work that on typical platforms should
not be done in the main thread or work queue. And in fact, clients of WebKit like Safari
on iOS implement this asynchronously. But on platforms like Mac where the implementation
inside WebKit is used, we were doing the work synchronously. We&apos;ve seen hang reports from
Mac Safari that show us this is a problem for some users in practice. The fix is to use
a separate work queue for this, analogous to what clients were already doing.

* Source/WebCore/Sources.txt: Added SerializedCryptoKeyWrap.cpp.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Ditto.

* Source/WebCore/crypto/SerializedCryptoKeyWrap.cpp: Added.
(WebCore::getDefaultWebCryptoMasterKey): Added. Calls defaultWebCryptoMasterKey on
a dedicated work queue and then marshals the result back to the main work queue.
* Source/WebCore/crypto/SerializedCryptoKeyWrap.h: Added getDefaultWebCryptoMasterKey.

* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCFB.cpp: Added missing include,
needed due to unified source differences.
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp: Ditto.
* Source/WebCore/crypto/cocoa/CryptoAlgorithmRSA_PSSMac.cpp: Ditto.
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp: Ditto.

* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageNavigationClient): Use getDefaultWebCryptoMasterKey.
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::legacyWebCryptoMasterKey): Ditto.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::getWebCryptoMasterKey): Ditto.

Canonical link: <a href="https://commits.webkit.org/297317@main">https://commits.webkit.org/297317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc17a9dc263190443f1bf51682bc64260714ae4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111318 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30984 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21449 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117350 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61586 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113280 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39566 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84606 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114265 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100213 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65052 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24623 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18353 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61170 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94663 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18420 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120525 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38367 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28507 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93536 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38743 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96487 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93360 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23790 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38453 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16225 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34369 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38256 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43733 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37921 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39623 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/crypto/SerializedCryptoKeyWrap.cpp:
```diff
@@ -0,0 +1,44 @@
+
+void getDefaultWebCryptoMasterKey(CompletionHandler<void(std::optional<Vector<uint8_t>>&&)>&& handler)
+{
+    static NeverDestroyed<Ref<WorkQueue>> queue { WorkQueue::create("org.WebKit.WebCryptoMasterKey"_s) };
```
Doing a quick a grep, it looks like we aren't too consistent with our work queue label naming. I see:

"org.webkit.foo"
"org.WebKit.foo"
"com.apple.WebKit.foo"
"WebKit::Foo Queue"
"Foo Data Queue"
"Foo Queue"

The most common seems to be "com.apple.WebKit.foo", but I like "org.WebKit.foo" more, so this seems like the right direction. Good place to cleanup.

darinadler replied:
I copied the idiom from `BundleResourceLoader.mm`. It didn’t even occur to me that we weren’t consistent!

In terms of what pattern to follow in future:

1. `org.webkit` is better than `com.apple.webkit`
2. queue name strings don’t need the word "queue" in them
3. identifying the project is important, will be confusing if threads are just named something like `WebCryptoMasterKey`

Not sure if:

1. `org.WebKit` is better than `org.webkit`
2. `org.WebKit` is better than `WebKit` or `WebKit::`
3. A dot-separated identifier is better than something with spaces

weinig commented on Source/WebCore/crypto/SerializedCryptoKeyWrap.h:
```diff
@@ -40,6 +40,7 @@ struct WrappedCryptoKey;
 // https://www.w3.org/TR/WebCryptoAPI/#security-developers
 
 WEBCORE_EXPORT std::optional<Vector<uint8_t>> defaultWebCryptoMasterKey();
+WEBCORE_EXPORT void getDefaultWebCryptoMasterKey(CompletionHandler<void(std::optional<Vector<uint8_t>>&&)>&&);
```
Is a get prefix the usual choice for async functions these days?

darinadler replied:
I think so. I found about 600 examples when I searched for `\bget[A-Z].*CompletionHandler` in WebKit.

weinig reviewed (approved):
One thing I am not entirely clear on is if this is an observable change that can be tested in an API test or not. I certainly don't have a clear picture in my head how I would test this.

darinadler committed:
Avoidable hangs in WebCore::defaultWebCryptoMasterKey
https://bugs.webkit.org/show_bug.cgi?id=295844
rdar://151058747

Reviewed by Sam Weinig.

The code in WebCore::defaultWebCryptoMasterKey does work that on typical platforms should
not be done in the main thread or work queue. And in fact, clients of WebKit like Safari
on iOS implement this asynchronously. But on platforms like Mac where the implementation
inside WebKit is used, we were doing the work synchronously. We've seen hang reports from
Mac Safari that show us this is a problem for some users in practice. The fix is to use
a separate work queue for this, analogous to what clients were already doing.

* Source/WebCore/Sources.txt: Added SerializedCryptoKeyWrap.cpp.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Ditto.

* Source/WebCore/crypto/SerializedCryptoKeyWrap.cpp: Added.
(WebCore::getDefaultWebCryptoMasterKey): Added. Calls defaultWebCryptoMasterKey on
a dedicated work queue and then marshals the result back to the main work queue.
* Source/WebCore/crypto/SerializedCryptoKeyWrap.h: Added getDefaultWebCryptoMasterKey.

* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESCFB.cpp: Added missing include,
needed due to unified source differences.
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp: Ditto.
* Source/WebCore/crypto/cocoa/CryptoAlgorithmRSA_PSSMac.cpp: Ditto.
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp: Ditto.

* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageNavigationClient): Use getDefaultWebCryptoMasterKey.
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::legacyWebCryptoMasterKey): Ditto.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::getWebCryptoMasterKey): Ditto.

Canonical link: https://commits.webkit.org/297317@main
